packages/core/src/utils/request.ts

```diff
@@ -0,0 +1,135 @@
+  const headers: Record<string, string> = {};
+  try {
+    winterCGHeaders.forEach((value, key) => {
+      if (typeof value === 'string') {
+        // We check that value is a string even though it might be redundant to make sure prototype pollution is not possible.
+        headers[key] = value;
+      }
+    });
+  } catch {
+    // just return the empty headers
+  }
+
+  return headers;
+}
+
+/**
+ * Convert common request headers to a simple dictionary.
+ */
+export function headersToDict(reqHeaders: Record<string, string | string[] | undefined>): Record<string, string> {
+  const headers: Record<string, string> = Object.create(null);
+
+  try {
+    Object.entries(reqHeaders).forEach(([key, value]) => {
+      if (typeof value === 'string') {
+        headers[key] = value;
+      }
+    });
+  } catch {
+    // just return the empty headers
+  }
+
+  return headers;
+}
+
+/**
+ * Converts a `Request` object that implements the `Web Fetch API` (https://developer.mozilla.org/en-US/docs/Web/API/Headers) into the format that the `RequestData` integration understands.
+ */
+export function winterCGRequestToRequestData(req: WebFetchRequest): RequestEventData {
+  const headers = winterCGHeadersToDict(req.headers);
+
+  return {
+    method: req.method,
+    url: req.url,
+    query_string: extractQueryParamsFromUrl(req.url),
+    headers,
+    // TODO: Can we extract body data from the request?
+  };
+}
+
+/**
+ * Convert a HTTP request object to RequestEventData to be passed as normalizedRequest.
+ * Instead of allowing `PolymorphicRequest` to be passed,
+ * we want to be more specific and generally require a http.IncomingMessage-like object.
+ */
+export function httpRequestToRequestData(request: {
+  method?: string;
+  url?: string;
+  headers?: {
+    [key: string]: string | string[] | undefined;
+  };
+  protocol?: string;
+  socket?: {
+    encrypted?: boolean;
+    remoteAddress?: string;
+  };
+}): RequestEventData {
+  const headers = request.headers || {};
+  const host = typeof headers.host === 'string' ? headers.host : undefined;
+  const protocol = request.protocol || (request.socket?.encrypted ? 'https' : 'http');
+  const url = request.url || '';
+
+  const absoluteUrl = getAbsoluteUrl({
+    url,
+    host,
+    protocol,
+  });
+
+  // This is non-standard, but may be sometimes set
+  // It may be overwritten later by our own body handling
+  const data = (request as PolymorphicRequest).body || undefined;
+
+  // This is non-standard, but may be set on e.g. Next.js or Express requests
+  const cookies = (request as PolymorphicRequest).cookies;
+
+  return dropUndefinedKeys({
+    url: absoluteUrl,
+    method: request.method,
+    query_string: extractQueryParamsFromUrl(url),
+    headers: headersToDict(headers),
+    cookies,
+    data,
+  });
+}
+
+function getAbsoluteUrl({
+  url,
+  protocol,
+  host,
+}: { url?: string; protocol: string; host?: string }): string | undefined {
+  if (url && url.startsWith('http')) {
```

Comment:
previously, we checked here `url.startsWith(protocol)`. Which should be generally fine, but there could be an edge case where you have `url: http://example.com` and `protocol: https` which would result in `https://http://example.com` which seems bad.